### PR TITLE
Migrate part of s3 tests to dynamic env

### PR DESCRIPTION
**Pull request by vvarg229**

### CODE REVIEW

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+    get_file_hash,
+    split_file,
+)
+from s3_helper import (
+    check_objects_in_bucket,
+    check_tags_by_bucket,
+    check_tags_by_object,
+    set_bucket_versioning,
+    try_to_get_objects_and_expect_error,
+    parametrize_clients
```
Comment (cthulhu-rider) [inline dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:22]: shouldnt it be `parametErize`?

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+from s3.s3_gate_base import TestNeofsS3GateBase
+
+logger = logging.getLogger("NeoLogger")
+
+
+def pytest_generate_tests(metafunc):
+    parametrize_clients(metafunc)
+
+
+@allure.link("https://github.com/nspcc-dev/neofs-s3-gw#neofs-s3-gateway", name="neofs-s3-gateway")
```
Comment (cthulhu-rider) [inline dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:35]: to me its too hard ref for this code space, i'd place it in some top-level header file

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+            buckets = s3_gate_bucket.list_buckets_s3(self.s3_client)
+            assert bucket_1 in buckets, f"Expected bucket {bucket_1} is in the list"
+            assert bucket_2 in buckets, f"Expected bucket {bucket_2} is in the list"
+
+        with allure.step("Bucket must be empty"):
+            for bucket in (bucket_1, bucket_2):
+                objects_list = s3_gate_object.list_objects_s3(self.s3_client, bucket)
+                assert not objects_list, f"Expected empty bucket, got {objects_list}"
+
+        with allure.step("Check buckets are visible with S3 head command"):
```
Comment (cthulhu-rider) [inline dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:63]:
```suggestion
        with allure.step("Check buckets are visible with S3 API HeadBucket op"):
```

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+            s3_gate_bucket.head_bucket(self.s3_client, bucket_1)
```
Comment (cthulhu-rider) [inline dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:64]: im newbee in Python functionality. IIUC if this instruction throws any exception, the test step will fail, right?

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+
+        self.s3_client.sync(bucket_name=bucket, dir_path=os.path.dirname(file_path_1))
+
+        with allure.step("Check objects are synced"):
+            objects = s3_gate_object.list_objects_s3(self.s3_client, bucket)
+
+        with allure.step("Check these are the same objects"):
+            assert set(key_to_path.keys()) == set(
+                objects
+            ), f"Expected all objects saved. Got {objects}"
```
Comment (cthulhu-rider) [inline dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:160]: it seems to me that will we see this failure also then `objects` set is wider then `key_to_path.keys()`

if so, i'd rephrase
```suggestion
            ), f"Expected exact objects saved. Got {objects}"
```

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+        assert not uploads, f"Expected there is no uploads in bucket {bucket}"
+
+        with allure.step("Create and abort multipart upload"):
+            upload_id = s3_gate_object.create_multipart_upload_s3(
+                self.s3_client, bucket, object_key
+            )
+            uploads = s3_gate_object.list_multipart_uploads_s3(self.s3_client, bucket)
+            assert uploads, f"Expected there one upload in bucket {bucket}"
+            assert (
+                uploads[0].get("Key") == object_key
```
Comment (cthulhu-rider) [inline dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:273]: not related to this exact test, but i suddenly became curious :thinking:

tested cluster is a single connected node system with a fixed synchronization model. It would be extremely useful to test the behavior of paired operations (like here `create->list`) accessing different segments of the cluster. I see this as a separate class of tests, but which is no less critical for cluster arch's

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+        key_value_pair_bucket = [("some-key", "some-value"), ("some-key-2", "some-value-2")]
+        key_value_pair_obj = [
+            ("some-key-obj", "some-value-obj"),
+            ("some-key--obj2", "some-value--obj2"),
+        ]
+        key_value_pair_obj_new = [("some-key-obj-new", "some-value-obj-new")]
+        file_name_simple = generate_file(simple_object_size)
+        obj_key = self.object_key_from_file_path(file_name_simple)
+
+        s3_gate_bucket.put_bucket_tagging(self.s3_client, bucket, key_value_pair_bucket)
```
Comment (cthulhu-rider) [inline dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:339]: are we obliged by S3 protocol to `PutBucketTagging` before `PutObjectTagging`?

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+                tags,
+            )
+
+        s3_gate_object.delete_object_tagging(self.s3_client, bucket, obj_key)
+        check_tags_by_object(self.s3_client, bucket, obj_key, [])
+
+    @allure.title("Test S3: Delete object & delete objects S3 API")
+    def test_s3_api_delete(self, two_buckets, simple_object_size, complex_object_size):
+        """
+        Check delete_object and delete_objects S3 API operation. From first bucket some objects deleted one by one.
```
Comment (cthulhu-rider) [inline dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:358]:
```suggestion
        Check DeleteObject and DeleteObjects S3 API operations. From first bucket some objects deleted one by one.
```

**Review by cthulhu-rider — COMMENTED** [review]
@vvarg229 i've reviewed `bucket` and `gate` files only by now. Overall everything looks great to me

it's a bit strange to me to see `test_s3_gate.py` file in `s3_gate` besides seeing S3 API tests in it. Is this file intended for NeoFS S3 Gateway specifics not covered by the protocol?

i've left some questions and suggestions from my own, if u'll find any of them useful, feel free to postpone them via issues, i understand that this PR is about migration, not enhancements

i don't plan to look closely at the other tests (at least for now), hope @evgeniiz321 @roman-khimov will help

**Review by evgeniiz321 — APPROVED** [review]

**Review by MaxGelbakhiani — APPROVED** [review]

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_bucket.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+            bucket_acl_3 = s3_gate_bucket.get_bucket_acl(self.s3_client, bucket_3)
+            assert_bucket_s3_acl(
+                acl_grants=bucket_acl_3, permitted_users="AllUsers", acl=acl
+            )
+
+    @pytest.mark.acl
+    @allure.title("Test S3: Create Bucket with different ACL by grand")
+    def test_s3_create_bucket_with_grands(self):
+
+        with allure.step("Create bucket with  --grant-read"):
```
[the remark this answers is not in the document]
Comment (vvarg229) [reply dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_bucket.py:69]: https://github.com/nspcc-dev/neofs-testcases/issues/701

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_bucket.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+@pytest.mark.s3_gate_bucket
+class TestS3GateBucket(TestNeofsS3GateBase):
+    @pytest.mark.acl
+    @pytest.mark.sanity
+    @allure.title("Test S3: Create Bucket with different ACL")
+    def test_s3_create_bucket_with_ACL(self):
+
+        with allure.step("Create bucket with ACL private"):
+            acl="private"
+            bucket = s3_gate_bucket.create_bucket_s3(self.s3_client, True, acl=acl, bucket_configuration="rep-1")
```
[the remark this answers is not in the document]
Comment (vvarg229) [reply dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_bucket.py:31]: > i can only imagine what parameter is set to `True` (or `False` in other cases). Suggest to create named const/var (idk whats better in Python)

https://github.com/nspcc-dev/neofs-testcases/pull/702

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_bucket.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    if "s3_client" in metafunc.fixturenames:
+        metafunc.parametrize("s3_client", ["aws cli", "boto3"], indirect=True)
+
+
+@pytest.mark.s3_gate
+@pytest.mark.s3_gate_bucket
+class TestS3GateBucket(TestNeofsS3GateBase):
+    @pytest.mark.acl
+    @pytest.mark.sanity
+    @allure.title("Test S3: Create Bucket with different ACL")
```
[the remark this answers is not in the document]
Comment (vvarg229) [reply dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_bucket.py:26]: https://github.com/nspcc-dev/neofs-testcases/pull/703

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_bucket.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        with allure.step("Put two objects into bucket"):
+            s3_gate_object.put_object_s3(self.s3_client, bucket, file_path_1)
+            s3_gate_object.put_object_s3(self.s3_client, bucket, file_path_2)
+            check_objects_in_bucket(self.s3_client, bucket, [file_name_1, file_name_2])
+
+        with allure.step("Try to delete not empty bucket and get error"):
+            with pytest.raises(Exception, match=r".*The bucket you tried to delete is not empty.*"):
+                s3_gate_bucket.delete_bucket_s3(self.s3_client, bucket)
+
+        with allure.step("Delete object in bucket"):
```
[the remark this answers is not in the document]
Comment (vvarg229) [reply dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_bucket.py:157]: https://github.com/nspcc-dev/neofs-testcases/pull/703

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py (hunk 3 above)
Comment (vvarg229) [reply dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:63]: https://github.com/nspcc-dev/neofs-testcases/pull/703

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py (hunk 5 above)
Comment (vvarg229) [reply dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:160]: https://github.com/nspcc-dev/neofs-testcases/pull/703

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py (hunk 4 above)
Comment (vvarg229) [reply dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:64]: > im newbee in Python functionality. IIUC if this instruction throws any exception, the test step will fail, right?

Yes, it will

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_bucket.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+                True,
+                grant_read="uri=http://acs.amazonaws.com/groups/global/AllUsers",
+                bucket_configuration="rep-1",
+            )
+            bucket_acl = s3_gate_bucket.get_bucket_acl(self.s3_client, bucket)
+            assert_bucket_s3_acl(
+                acl_grants=bucket_acl, permitted_users="AllUsers", acl="grant-read"
+            )
+
+        with allure.step("Create bucket with --grant-wtite"):
```
[the remark this answers is not in the document]
Comment (vvarg229) [reply dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_bucket.py:81]: https://github.com/nspcc-dev/neofs-testcases/pull/703

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py (hunk 7 above)
Comment (vvarg229) [reply dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:339]: > are we obliged by S3 protocol to `PutBucketTagging` before `PutObjectTagging`?

As I understand from the documentation - no, we don't have to use `PutBucketTagging` before `PutObjectTagging`.

I tested it - you are right, the tests run the same way.

See PR:
https://github.com/nspcc-dev/neofs-testcases/pull/705

File: dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py (hunk 6 above)
Comment (vvarg229) [reply dynamic_env_pytest_tests/tests/services/s3_gate/test_s3_gate.py:273]: https://github.com/nspcc-dev/neofs-testcases/issues/706
